server/src/main/java/org/opensearch/gateway/remote/routingtable/RemoteIndexRoutingTableDiff.java

```diff
@@ -0,0 +1,134 @@
+import org.opensearch.core.common.io.stream.BufferedChecksumStreamInput;
+import org.opensearch.core.common.io.stream.BufferedChecksumStreamOutput;
+import org.opensearch.core.common.io.stream.InputStreamStreamInput;
+import org.opensearch.core.common.io.stream.StreamOutput;
+import org.opensearch.core.common.io.stream.Writeable;
+
+import java.io.EOFException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.HashMap;
+import java.util.Map;
+
+public class RemoteIndexRoutingTableDiff implements Diff<IndexRoutingTable>, Writeable {
+
+    private final Map<String, Diff<IndexRoutingTable>> diffs;
+    private final String codec = "RemoteIndexRoutingTableDiff";
+
+    public static final int VERSION = 1;
+
+    /**
+     * Constructs a new RemoteIndexRoutingTableDiff with the given differences.
+     *
+     * @param diffs a map containing the differences of {@link IndexRoutingTable}.
+     */
+    public RemoteIndexRoutingTableDiff(Map<String, Diff<IndexRoutingTable>> diffs) {
+        this.diffs = diffs;
+    }
+
+    /**
+     * Reads data from inputStream and creates RemoteIndexRoutingTableDiff object with the {@link Diff<IndexRoutingTable>}
+     *
+     * @param inputStream input stream with diff data
+     * @throws IOException exception thrown on failing to read from stream.
+     */
+    public RemoteIndexRoutingTableDiff(InputStream inputStream) throws IOException {
+        try (BufferedChecksumStreamInput in = new BufferedChecksumStreamInput(new InputStreamStreamInput(inputStream), "assertion")) {
+            CodecUtil.checkHeader(new InputStreamDataInput(inputStream), codec, VERSION, VERSION);
+            int size = in.readVInt();
+            Map<String, Diff<IndexRoutingTable>> diffs = new HashMap<>();
+            for (int i = 0; i < size; i++) {
+                String key = in.readString();
+                Diff<IndexRoutingTable> diff = IndexRoutingTable.readDiffFrom(in);
+                diffs.put(key, diff);
+            }
+            verifyCheckSum(in);
+            this.diffs = diffs;
+        } catch (EOFException e) {
+            throw new IOException("Diffs data is corrupted", e);
+        }
+    }
+
+    /**
+     * Writes {@link Diff<IndexRoutingTable>} to the given stream
+     *
+     * @param streamOutput output stream to write
+     * @throws IOException exception thrown on failing to write to stream.
+     */
+    @Override
+    public void writeTo(StreamOutput streamOutput) throws IOException {
+        try (BufferedChecksumStreamOutput out = new BufferedChecksumStreamOutput(streamOutput)) {
+            // Write codec header
+            CodecUtil.writeHeader(new OutputStreamDataOutput(streamOutput), codec, VERSION);
+
+            out.writeVInt(diffs.size());
+            for (Map.Entry<String, Diff<IndexRoutingTable>> entry : diffs.entrySet()) {
+                out.writeString(entry.getKey());
+                entry.getValue().writeTo(out);
+            }
```

Comment:
It would be Map<String, List<IndexShardRoutingTable>>
Sample
```
value = {IndexRoutingTableDiff@14946} 
   indexShardRoutingTables = {ArrayList@14947}  size = 1
    0 = {IndexShardRoutingTable@14949} "IndexShardRoutingTable([test-index][5]){[test-index][5], node[epWnFAyDRh6PPV8xlo7InQ], [P], s[STARTED], a[id=KjBzDynrR4uSeLgLMMy8jQ], [test-index][5], node[null], [R], recovery_source[peer recovery], s[UNASSIGNED], unassigned_info[[reason=INDEX_CREATED], at[2024-07-10T10:09:09.695Z], delayed=false, allocation_status[no_attempt]]}"
  key = "test-index"
```